### PR TITLE
Arthur - JS Bowling Challenge

### PR DESCRIPTION
**Pull request by arthurfincham**
Hope you like it! The code might be a bit confusing in places so let me know if anything doesn't make sense..

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,60 @@
...
+
+<h2>Preview</h2>
+
+<img src="images/preview.gif" width="400px">
+
+_command line interface for the game_
+
+</div>
+
+## Install
```
Comment (AJOsmaston) [inline README.md:18]: This is a high quality readme. Was very easy for me to set up the project and test it out myself.

File: spec/final_frame.test.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+const { describe } = require('jest-circus');
+const FinalFrame = require('../src/final_frame');
+
+describe('final frame', () => {
```
Comment (AJOsmaston) [inline spec/final_frame.test.js:4]: Would you have wanted to add tests to third roll here?

File: spec/final_frame.test.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  const frame = new FinalFrame();
+
+  test('sets rolls at null', () => {
+    expect(frame.first_roll).toEqual(null);
+    expect(frame.second_roll).toEqual(null);
+    expect(frame.third_roll).toEqual(null);
```
Comment (AJOsmaston) [inline spec/final_frame.test.js:10]: I mean, in addition to this one

File: spec/game.test.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
+const { describe } = require('jest-circus');
+const Game = require('../src/game');
+
+beforeEach(() => {
+  game = new Game();
+});
+
+describe('Game', () => {
+  test('has an empty array of frames', () => {
+    expect(game.frames).toEqual([]);
```
Comment (AJOsmaston) [inline spec/game.test.js:10]: I don't think tests like these are overly helpful - if you wanted to change game.frames to be a frame object instead of an array, even if the program still works the way you want it to (ie the behaviour is the same) the test will now fail.

File: spec/game.test.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+});
+
+describe('Spares', () => {
+  test('spare returns true', () => {
+    const mockFrame = { first_roll: 4, second_roll: 6, score: 10 };
+    expect(game.isSpare(mockFrame)).toEqual(true);
+  });
+});
+
+describe('if first_roll exceeds 10', () => {
```
Comment (AJOsmaston) [inline spec/game.test.js:39]: I like that you've included tests like these in order to catch exceptions when playing in the command line.

**Review by AJOsmaston — COMMENTED** [review]
Overall this is a great project, coupled with an amazing README. Honestly great job.

The code is elegant - classes are distinct and have single responsibilities, functions are similarly succinct and easy to follow. 

The tests are thorough - I like your inclusion of testing for edge cases, which I think is necessary with the inclusion of the command line interface, as you can no longer assume a standardised input. I do think your tests sometimes err on the side of being state tests - if you can change how your functions work while retaining the overall functionality, but the tests now fail, they were more than likely testing the state rather than the overall behaviour.

Finally, a couple of ideas for future expansion - in RL bowling, the scorecard will overwrite the first 10 with an x (rather than the second space), and a spare will have the second digit overwritten with a '/'. The score will also display cumulatively, rather than an individual score for each frame - and not display until the score is ready to be calculated. 

have a look at :
https://www.bowlinggenius.com/
for an example of what I'm talking about.

I am, however, nitpicking - this is, overall, a really great project and was a pleasure to look over.

File: spec/game.test.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    game = new Game();
+
+    multiRoll = (num, one, two) => {
+      for (let i = 0; i < num; i += 1) {
+        game.roll(one, two);
+      }
+    };
+  });
+
+  test('returns 0 for a gutter game', () => {
```
Comment (eoinmakers) [inline spec/game.test.js:78]: Comprehensive testing

File: spec/game.test.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+describe('Game', () => {
+  test('adds a frame to the array', () => {
+    const mockFrame = { first_roll: null, second_roll: null, score: 0 };
+    game.newFrame(mockFrame);
+    expect(game.frames.length).toEqual(1);
+  });
+});
+
+describe('Strikes', () => {
```
Comment (eoinmakers) [inline spec/game.test.js:16]: Clear organisation of your unit tests.

File: spec/game.test.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  });
+
+  test('returns 150 for a game of fives', () => {
+    multiRoll(9, 5, 5);
+    game.finalRoll(5, 5);
+    game.bonusRoll(5);
+    expect(game.fetchScore()).toEqual(150);
+  });
+
+  test('returns 300 for a perfect game', () => {
```
Comment (eoinmakers) [inline spec/game.test.js:97]: Nice work.

File: src/game.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+  roll(one, two) {
+    this.rollParams(one, two);
+    this.newFrame(one, two);
+    if (this.frames.length > 2) {
+      this.bonusScan();
+    }
+  }
+
+  finalRoll(one, two) {
```
Comment (eoinmakers) [inline src/game.js:18]: Slightly better method names align with an action or verb describing what it's doing, eg. runFinalRoll, calculateFinalRoll, isFinalRoll, etc.

**Review by eoinmakers — COMMENTED** [review]
Thought this was a well written bowling challenge - just the note around method naming is the main feedback here, as it can be applied to refactoring each class. Nice work!
